### ISSUE TITLE
Using `Space` as a derived prop is terrible for re-rendering performance.

### ISSUE THREAD

**Issue by fasterthanlime** [itchio/itch#2054]
This is a React/Redux thing.

Once upon a time, I was doing the hardcore React/Redux thing, which was to:

  * Have everything in a state tree
  * Use `connect` to pick the bits of the part you needed
  * Let react-redux (and reselect) do `shouldComponentUpdate` for you
  * And that makes sure that components only re-render when they actually need to

And then it got really annoying, and somewhere around the time I started doing butlerd calls directly from the renderer I introduced the `withSpace` higher-order-component, which

  * Is built from a `TabInstance`
  * contains things like.. the parsed URL, whether a tab is loading, etc.
    * tl;dr it extracts parts of the tab we need.

### Why is Space good?

Here's an example of Space usage in `BrowserContext/index.tsx`:

```typescript
class BrowserContext extends React.PureComponent<Props> {
  render() {
    const { space } = this.props;
    if (space.prefix === "games") {
      const gameId = space.numericId();
      return (
        <BrowserContextContainer>
          <FetchGame
            params={{ gameId }}
            sequence={space.sequence()}
            render={({ result }) => {
              return <BrowserContextGame game={result.game} />;
            }}
          />
        </BrowserContextContainer>
      );
    }

    return null;
  }
}

interface Props {
  space: Space;
  dispatch: Dispatch;
}

export default withSpace(hook()(BrowserContext));
```

Looks good-ish right? We're actually only using three bits of the TabInstance here:

  * The space's `sequence` (that gets incremented when we reload a tab)
  * The space's `prefix` (which is the `games` part in `itch://games/3`)
  * The space's `numericId` (the `3`, as a number, in `itch://games/3`)

Note that only `sequence` is explicitly a property of `TabInstance`:

```typescript
export interface TabInstance {
  /** pages visited in this tab */
  history: TabPage[];

  /** current index of history shown */
  currentIndex: number;

  loading?: boolean;

  /** if sleepy, don't load until it's focused */
  sleepy?: boolean;

  /** label we had when saving the tab */
  savedLabel?: LocalizedString;

  /** number that increments when we reload a tab */
  sequence: number;
}
```

...the rest is parsed from the URL, which is `tabInstance.history[tabInstance.currentIndex].url`.

A cool thing about Space is that it has these methods to, for example make actions we want to dispatch. So if we wanted to update the label of the tab we're showing a view in:

```typescript
  componentDidMount() {
    const { space, dispatch } = this.props;
    dispatch(
      space.makePageUpdate({
        label: ["sidebar.applog"],
      })
    );
    this.queueFetch();
  }
```

...which is very convenient.

### Why is Space bad?

A React component re-renders:

  * When you call `component.forceUpdate()` (which we never do)
  * When
    * a component's state changes (via `component.setState()`) or a component's props changes (via a parent re-rendering, with the same child, but with different props)
    * *and* `shouldComponentUpdate` returns true

When we use `connect` (above shown as `hook`, which is our own little spin on it that has better type safety), bits of the Redux state are flattened into individual props, and since `shouldComponentUpdate` does a shallow-compare between every prop (more or less), most useless renders are prevented.

But when we use Space, well.. here's the hoc:

```typescript
import React from "react";
import { Space } from "common/helpers/space";
import { Subtract } from "common/types";

export interface SpaceContextProps {
  space: Space;
}

const spaceContext = React.createContext<Space>(undefined);
export const SpaceProvider = spaceContext.Provider;
export const SpaceConsumer = spaceContext.Consumer;

export const withSpace = <P extends SpaceContextProps>(
  Component: React.ComponentType<P>
) => (props: Subtract<P, SpaceContextProps>) => (
  <SpaceConsumer>
    {space => <Component {...props} space={space} />}
  </SpaceConsumer>
);
```

it just injects the `space` prop. Which it gets from wherever `<SpaceProvider/>` is used, which happens to be here:

```typescript
    return (
      <MeatContainer onClick={this.onClick}>
        <TitleBar tab={currentId} />
        {map(openTabs, tab => {
          const tabInstance = tabInstances[tab];
          const space = Space.fromInstance(tab, tabInstance);
          const visible = tab === currentId;
          return (
            <SpaceProvider key={tab} value={space}>
              <MeatTab
                key={tab}
                data-id={tab}
                data-url={space.url()}
                data-resource={space.resource()}
                className={classNames("meat-tab", { visible })}
              >
                <Meat visible={visible} sequence={tabInstance.sequence} />
              </MeatTab>
            </SpaceProvider>
          );
        })}
      </MeatContainer>
    );
```

...so the problem here is that every time `tabInstances` changes (which is quite often as it turns out!), we make a new `Space` instance, so *every components that uses `withSpace` re-renders*.

And that turns out to be rather costly. For example, here's what hitting refresh on the "Owned items" view does:

![image](https://user-images.githubusercontent.com/7998310/44537309-b72ec600-a6fe-11e8-9ff4-f6993c8104e2.png)

In more details:

![image](https://user-images.githubusercontent.com/7998310/44537331-c6157880-a6fe-11e8-90f3-d26bbbd830ee.png)

As javascript function calls instead of User Timings:

![image](https://user-images.githubusercontent.com/7998310/44537348-d62d5800-a6fe-11e8-96ed-11cf428a117f.png)

Note that *nothing on the screen changed*, and yet we still re-rendered everything, which took about 160ms when I ran this test (in development, so, much slower than it would be in production).

It gets worse: every change to a tab makes every other tab re-render, and that gets costly *fast* if you have a bunch of native views in the background.

### How do we fix it?

The way I see it, Space has to die - at least in the renderer. All of Space's methods should be denormalized in the reducer instead.

Then, components should stop using `withSpace` and just use `hook` and extract the properties they need. Luckily, this can be done gradually.

**Comment by fasterthanlime** [comment]
I'll track tasks here:

  * [x] Make the `tab-instance` reducer work the same way `wind` reducer does:
    * ie. separate the reducer that handles the `tabInstances` map, from the reducer that handles each individual `tabInstance` - this makes writing the latter much easier and less error-prone.
  * [x] Have the individual `tabInstance` reducer parse the URL (when it changes!) and flatten properties like `protocol`, `prefix`, `numericId`, query parameters, but also stuff like `canGoForward`, `canGoBackward`, `isBrowser`, etc.
  * [x] Switch one component from `withSpace` to regular `hook`
  * [x] Confirm the effects on User Timings
  * [x] Gradually change the others
  * [x] Eventually remove `withSpace`

**Comment by fasterthanlime** [comment]
Here's a good example: changing the filter in the dashboard:

![update](https://user-images.githubusercontent.com/7998310/44540336-ca459400-a706-11e8-902a-ed954e71088f.png)

Everything updates a first time (when `sequence` is changed), and then a second time (when the `Fetch.ProfileGames` request returns). Note that the request itself takes 7ms! It could fit within a frame.
